// ethers/src.ts/utils/index.ts
/**
 *  There are many simple utilities required to interact with
 *  Ethereum and to simplify the library, without increasing
 *  the library dependencies for simple functions.
 *
 *  @_section api/utils:Utilities  [about-utils]
 */

export { decodeBase58, encodeBase58 } from "ethers/src.ts/utils/base58";

export { decodeBase64, encodeBase64 } from "ethers/src.ts/utils/base64";

export {
    getBytes, getBytesCopy, isHexString, isBytesLike, hexlify, concat, dataLength, dataSlice,
    stripZerosLeft, zeroPadValue, zeroPadBytes
} from "ethers/src.ts/utils/data";

export {
    isCallException, isError,
    assert, assertArgument, assertArgumentCount, assertPrivate, assertNormalize, makeError
} from "ethers/src.ts/utils/errors"

export { EventPayload } from "ethers/src.ts/utils/events";

export {
    FetchRequest, FetchResponse, FetchCancelSignal,
} from "ethers/src.ts/utils/fetch";

export { FixedNumber } from "ethers/src.ts/utils/fixednumber"

export {
    fromTwos, toTwos, mask,
    getBigInt, getNumber, getUint, toBigInt, toNumber, toBeHex, toBeArray, toQuantity
} from "ethers/src.ts/utils/maths";

export { resolveProperties, defineProperties} from "ethers/src.ts/utils/properties";

export { decodeRlp } from "ethers/src.ts/utils/rlp-decode";
export { encodeRlp } from "ethers/src.ts/utils/rlp-encode";

export { formatEther, parseEther, formatUnits, parseUnits } from "ethers/src.ts/utils/units";

export {
    toUtf8Bytes,
    toUtf8CodePoints,
    toUtf8String,

    Utf8ErrorFuncs,
} from "ethers/src.ts/utils/utf8";

export { uuidV4 } from "ethers/src.ts/utils/uuid";

/////////////////////////////
// Types

export type { BytesLike } from "ethers/src.ts/utils/data";

export type {

    //ErrorFetchRequestWithBody, ErrorFetchRequest,
    //ErrorFetchResponseWithBody, ErrorFetchResponse,

    ErrorCode,

    EthersError, UnknownError, NotImplementedError, UnsupportedOperationError, NetworkError,
    ServerError, TimeoutError, BadDataError, CancelledError, BufferOverrunError,
    NumericFaultError, InvalidArgumentError, MissingArgumentError, UnexpectedArgumentError,
    CallExceptionError, InsufficientFundsError, NonceExpiredError, OffchainFaultError,
    ReplacementUnderpricedError, TransactionReplacedError, UnconfiguredNameError,
    ActionRejectedError,

    CallExceptionAction, CallExceptionTransaction,

    CodedEthersError
} from "ethers/src.ts/utils/errors"

export type { EventEmitterable, Listener } from "ethers/src.ts/utils/events";

export type {
    GetUrlResponse,
    FetchPreflightFunc, FetchProcessFunc, FetchRetryFunc,
    FetchGatewayFunc, FetchGetUrlFunc
} from "ethers/src.ts/utils/fetch";

export type { FixedFormat } from "ethers/src.ts/utils/fixednumber"

export type { BigNumberish, Numeric } from "ethers/src.ts/utils/maths";

export type { RlpStructuredData, RlpStructuredDataish } from "ethers/src.ts/utils/rlp";

export type {
    Utf8ErrorFunc,
    UnicodeNormalizationForm,
    Utf8ErrorReason
} from "ethers/src.ts/utils/utf8";
