// ethers/src.ts/abi/index.ts
/**
 *  The Application Binary Interface (ABI) describes how method input
 *  parameters should be encoded, their results decoded, and how to
 *  decode events and errors.
 *
 *  See [About ABIs](docs-abi) for more details how they are used.
 *
 *  @_section api/abi:Application Binary Interface  [about-abi]
 *  @_navTitle: ABI
 */


//////
export { AbiCoder } from "ethers/src.ts/abi/abi-coder";

export { decodeBytes32String, encodeBytes32String } from "ethers/src.ts/abi/bytes32";

export {
    ConstructorFragment, ErrorFragment, EventFragment, FallbackFragment,
    Fragment, FunctionFragment, NamedFragment, ParamType, StructFragment,
} from "ethers/src.ts/abi/fragments";

export {
    checkResultErrors,
    Indexed,
    Interface,
    ErrorDescription, LogDescription, TransactionDescription,
    Result
} from "ethers/src.ts/abi/interface";

export { Typed } from "ethers/src.ts/abi/typed";

export type {
    JsonFragment, JsonFragmentType,
    FormatType, FragmentType, ParamTypeWalkAsyncFunc, ParamTypeWalkFunc
} from "ethers/src.ts/abi/fragments";

export type {
    InterfaceAbi,
} from "ethers/src.ts/abi/interface";
