// ethers/src.ts/wallet/index.ts
/**
 *  When interacting with Ethereum, it is necessary to use a private
 *  key authenticate actions by signing a payload.
 *
 *  Wallets are the simplest way to expose the concept of an
 *  //Externally Owner Account// (EOA) as it wraps a private key
 *  and supports high-level methods to sign common types of interaction
 *  and send transactions.
 *
 *  The class most developers will want to use is [[Wallet]], which
 *  can load a private key directly or from any common wallet format.
 *
 *  The [[HDNodeWallet]] can be used when it is necessary to access
 *  low-level details of how an HD wallets are derived, exported
 *  or imported.
 *
 *  @_section: api/wallet:Wallets  [about-wallets]
 */

export { BaseWallet } from "ethers/src.ts/wallet/base-wallet";

export {
    defaultPath,

    getAccountPath, getIndexedAccountPath,

    HDNodeWallet,
    HDNodeVoidWallet,
} from "ethers/src.ts/wallet/hdwallet";

export { isCrowdsaleJson, decryptCrowdsaleJson } from "ethers/src.ts/wallet/json-crowdsale";

export {
    isKeystoreJson,
    decryptKeystoreJsonSync, decryptKeystoreJson,
    encryptKeystoreJson, encryptKeystoreJsonSync
} from "ethers/src.ts/wallet/json-keystore";

export { Mnemonic } from "ethers/src.ts/wallet/mnemonic";

export { Wallet } from "ethers/src.ts/wallet/wallet";


export type { CrowdsaleAccount } from "ethers/src.ts/wallet/json-crowdsale";
export type {
    KeystoreAccount, EncryptOptions
} from "ethers/src.ts/wallet/json-keystore"
