// ethers/src.ts/contract/index.ts
/**
 *  A **Contract** object is a meta-class (a class whose definition is
 *  defined at runtime), which communicates with a deployed smart contract
 *  on the blockchain and provides a simple JavaScript interface to call
 *  methods, send transaction, query historic logs and listen for its events.
 *
 *  @_section: api/contract:Contracts  [about-contracts]
 */
export {
    BaseContract, Contract
} from "ethers/src.ts/contract/contract";

export {
    ContractFactory
} from "ethers/src.ts/contract/factory";

export {
    ContractEventPayload, ContractUnknownEventPayload,
    ContractTransactionReceipt, ContractTransactionResponse,
    EventLog, UndecodedEventLog
} from "ethers/src.ts/contract/wrappers";

export type {
    BaseContractMethod, ConstantContractMethod,
    PostfixOverrides,
    ContractEvent, ContractEventArgs, ContractEventName,
    ContractDeployTransaction,
    ContractInterface, ContractMethod, ContractMethodArgs, ContractTransaction,
    DeferredTopicFilter, Overrides,
    WrappedFallback
} from "ethers/src.ts/contract/types";
